include/circt/Dialect/Calyx/CalyxLoweringUtils.h

```diff
@@ -89,41 +131,351 @@ class WhileOpInterface {
   T impl;
 };
 
-// Walks the control of this component, and appends source information for leaf
-// nodes. It also appends a position attribute that connects the source location
-// metadata to the corresponding control operation.
-WalkResult
-getCiderSourceLocationMetadata(calyx::ComponentOp component,
-                               SmallVectorImpl<Attribute> &sourceLocations);
+//===----------------------------------------------------------------------===//
+// Lowering state classes
+//===----------------------------------------------------------------------===//
 
-// Tries to match a constant value defined by op. If the match was
-// successful, returns true and binds the constant to 'value'. If unsuccessful,
-// the value is unmodified.
-bool matchConstantOp(Operation *op, APInt &value);
+// An interface for the Calyx component lowering state. The `Loop` type must
+// implement the LoopInterface; it will be used for several lowering patterns.
+template <typename Loop>
+class ComponentLoweringStateInterface {
```

Comment:
Do you think it makes sense to remove the templating from `ComponentLoweringStateInterface` and then have a separate (templated) state for loop lowering which uses the ComponentLoweringState?
```
template<typename Loop>
class LoopLoweringStateInterface {
    LoopLoweringStateInterface<Loop>(ComponentLoweringStateInterface* cls);
    ...
```
That would allow us to untangle loop handling with "basic" behaviour, make `ComponentLoweringStateInterface` non-templated (easier to pass around), and move its definitions to a .cpp file.

Reply:
Yeah I think that's a good idea. Let me play around with it.

Reply:
I've taken a slightly different approach, since it looks like the loop lowering state is relatively independent of the ComponentLoweringStateInterface. PTAL.